test/e2e/capv_quick_start_test.go

```diff
@@ -37,3 +52,187 @@ var _ = Describe("Cluster Creation using Cluster API quick-start test", func() {
+
+	AfterEach(func() {
+		cleanupSpecNamespace(namespace)
+	})
+
+	It("should create a cluster with storage policy", func() {
+		clusterName := fmt.Sprintf("cluster-%s", util.RandomString(6))
+		Expect(namespace).NotTo(BeNil())
+
+		ds := getDatastore()
+		if ds == nil {
+			Fail("unable to retrieve datastore")
+			return
+		}
+
+		restClient := rest.NewClient(vsphereClient.Client)
+		Expect(restClient.Login(ctx, userInfo)).NotTo(HaveOccurred())
+		tagsManager := tags.NewManager(restClient)
+
+		cat := tags.Category{Name: fmt.Sprintf("%s-cat", clusterName), Cardinality: "SINGLE"}
+		Byf("creating category: %s", cat.Name)
+		catID, err := tagsManager.CreateCategory(ctx, &cat)
+		Expect(err).NotTo(HaveOccurred())
+		cat.ID = catID
+
+		tag := tags.Tag{Name: fmt.Sprintf("%s-tag", clusterName), CategoryID: catID}
+		Byf("creating tag: %s", tag.Name)
+		tagID, err := tagsManager.CreateTag(ctx, &tag)
+		Expect(err).NotTo(HaveOccurred())
+		tag.ID = tagID
+
+		By("attaching tag to datastore")
+		Expect(tagsManager.AttachTag(ctx, tag.ID, ds.Reference())).NotTo(HaveOccurred())
+
+		// create a storage policy with tag and category
+		spName := fmt.Sprintf("%s-sp", clusterName)
+		spID := createStoragePolicy(spName, cat.Name, tag.Name)
+		if spID == nil {
+			Fail("unable to create storage policy")
+			return
+		}
+
+		// creating machine template first as ApplyClusterTemplateAndWait will wait for machines to become ready
+		By("creating vsphere machine template")
+		vsphereMachineTemplate := makeVsphereMachineTemplate(clusterName, namespace.Name)
+		vsphereMachineTemplate.Spec.Template.Spec.StoragePolicyName = spName
+		Expect(bootstrapClusterProxy.GetClient().Create(ctx, vsphereMachineTemplate)).ShouldNot(HaveOccurred())
+
+		By("creating a workload cluster")
+		configCluster := defaultConfigCluster(clusterName, namespace.Name)
+		configCluster.Flavor = StoragePolicyFlavor
+
+		_ = clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy:                 bootstrapClusterProxy,
+			ConfigCluster:                configCluster,
+			WaitForClusterIntervals:      e2eConfig.GetIntervals("", "wait-cluster"),
+			WaitForControlPlaneIntervals: e2eConfig.GetIntervals("", "wait-control-plane"),
+			WaitForMachineDeployments:    e2eConfig.GetIntervals("", "wait-worker-nodes"),
+		})
+
+		By("verifying storage policy is used by VMs")
+		pbmClient, err := pbm.NewClient(ctx, vsphereClient.Client)
+		Expect(err).NotTo(HaveOccurred())
+		var res []types.PbmServerObjectRef
+		if pbmClient != nil {
+			res, err = pbmClient.QueryAssociatedEntity(ctx, types.PbmProfileId{UniqueId: spID.UniqueId}, "virtualMachine")
+			Expect(err).NotTo(HaveOccurred())
+		}
+		Expect(len(res)).To(BeNumerically(">", 0))
+
+		// Delete storage policy, tag and category
+		if pbmClient != nil {
+			_, _ = pbmClient.DeleteProfile(ctx, []types.PbmProfileId{*spID})
+		}
+		_ = tagsManager.DeleteTag(ctx, &tag)
+		_ = tagsManager.DeleteCategory(ctx, &cat)
```

Comment:
Call them with defer to make sure they get run at the end of the method. Since there is a return after a failure.

Also, even if the `pbmClient` fails to exist, we should enforce deletion of the profile.

```golang 
..
defer tagsManager.DeleteTag(ctx, &tag)
defer tagsManager.DeleteCategory(ctx, &cat)
```